include/motor.hpp

```diff
@@ -0,0 +1,242 @@
+#ifndef MOTOR_H
+#define MOTOR_H
+
+#include "Arduino.h"
+#include <exception.h>
+
+#include "freertos/FreeRTOS.h"
+#include "freertos/task.h"
+#include "esp_log.h"
+//#include "freertos/semphr.h"
+
+/**
+ *  Motor
+ *   
+ *  position = mm
+ *  speed = mm/s
+ *  acceleration = mm/s^2
+ */
+
+#define DEBUG_CLIPPING              // Show debug messages when motions violating the machine 
+                                    // physics are commanded
+
+typedef struct {
+  float start;
+  float end;
+  float keepout; /*  Soft endstop preventing hard crashes in mm. Will be 
+                  *  subtracted twice from physicalTravel. Should be 
+                  *  sufficiently to completley drive clear from 
+                  *  homing switch */
+} MachineGeometry;
+
+#define MOVEMENT_TASK_FLAG_EXIT (1 << 0)
+#define MOVEMENT_TASK_FLAG_NEXT (1 << 1)
+
+// TODO - Change flags into a FreeRTOS Event Group which allows better synchronizing on Flag State
+#define MOTOR_FLAG_ENABLED (1 << 0)
+
+#define MOTOR_FLAG_WARNING (1 << 1)
+#define MOTOR_FLAG_ERROR (1 << 2)
+#define MOTOR_FLAG_FATAL (1 << 3)
+
+#define MOTOR_FLAG_AT_TARGET (1 << 4)
+#define MOTOR_FLAG_MOTION_ACTIVE (1 << 5)
+#define MOTOR_FLAG_HOMED (1 << 6)
+
+#define MOTOR_FLAG_OVERHEATING (1 << 7)
+#define MOTOR_FLAG_OVERFORCE (1 << 8)
+
+typedef enum {
+  INACTIVE = 0,
+  ACTIVE = 1,
+  HOMING = 2, 
+  ERROR = 3
+} MotorState;
+
+typedef enum {
+  // Retraction Emergency useful for Throat play, where a quick retraction to a safe position is required
+  RETRACT = 0, 
+
+  // De-energize Emergency useful for Vaginal/Anal play, where a retraction could cause pain.
+  // Instead the motor itself is de-energized so no Holding Torque is experienced.
+  DEENERGIZE = 1 
+} MotorEmergency
```

Comment:
It's not complete yet, but I'm working on adding Emergency Stop functionality.
I'm curious what you think of the approach. I think it's important to have different types of Emergency Stop, but I'm not sure how to safely have the user choose.

You definitely want different types of emergency behavior depending on the situation.

I also want to add a Dead-Mans Switch to the Motor directly. So the Motor can monitor a pin/switch, and if it ever is inactive, it forces a Emergency Stop. I really want this Emergency functionality to bypass all User Integration to ensure guaranteed reliability (I don't want an in-experienced user to somehow accidentally mess up the Emergency Switch hookup).

I'm primarily interested in the Dead-Man's Switch case for Throat play, where a Dead-Mans Switch is important due to suffocation/passing out risk. I'll likely build in a normal EStop button where you need to intentionally press the switch for normal Vaginal/Anal play scenarios.

Reply:
I have thought about that a lot. Having a solid safety built-in is really important for me. I like the idea of a configurable emergency behaviour determined by the use case. Like stop, safe retract (speed should be configurable as well) or deenergize. 

However, I would leave the exact implementation of the safe behaviour up to the user. Building that into the motor driver will need additional hardware not everyone has.

So far the safety concept relied heavily on the ability to get away from the machine. One user did a complete hazard analysis (to be found somewhere on the OSSM GitHub). It's not as easy as adding a stop button. 

Anyway, when including an emergency stop this should cut the power to the complete machine rather then relying on unsafe code execution.